### ISSUE TITLE
No throttling on resize event handlers.

### ISSUE THREAD

**Issue by danigulyas** [storybooks/storybook#1990]
**Issue**

When you resize windows fast, everything is laggy, and you feel like:
![image](https://user-images.githubusercontent.com/10425681/31287729-96218a76-aac2-11e7-84d6-689b54774ebf.png)

---

My take on this:

**Cause**

The [resize event handler](https://github.com/storybooks/storybook/blob/master/lib/ui/src/modules/ui/components/layout/index.js#L138) is not [throttled](https://lodash.com/docs/4.17.4#throttle).

**Solution**

```jsx
import { throttle } from 'lodash'; // You have it as dependency already.

  {/* ... component stuff */}
   componentDidMount() {
    window.addEventListener('resize', this.onResize);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.onResize);
  }

  onResize = throttle(this.handleOnResize, 200, { trailing: true });

  {/* onResize renamed to handleOnResize */}
  {/* ... more component stuff */}
```

**Apart from that**

_Thank you for making this!_

<3
-Dan

**Comment by shilman** [comment]
@danigulyas care to issue a PR for this? it looks like a very reasonable solution, though I haven't run into the problem myself AFAIK so it's hard for me to tell!

**Comment by Hypnosphi** [comment]
I'd prefer to use `requestAnimationFrame` throttling rather than timer-based one

**Comment by danigulyas** [comment]
@shilman sure thing, somewhere on the weekend.
@Hypnosphi good idea, will check up on that!

**Comment by danigulyas** [comment]
So i dug into this more, did a little timeline action, turns out the resize is kind of slow in general, i've tested the same html inside and outside the context of Storybook, it was similarly slow, which is in a way understandable for a while.

However, after taking a second look on `onResize`, it seems to me that it couldn't even react to events (its just a higher order function, should've seen that before opening the issue).

See, `onResize` is registered as an event handler [here](https://github.com/storybooks/storybook/blob/master/lib/ui/src/modules/ui/components/layout/index.js#L138), but it is actually [being used for something different](https://github.com/storybooks/storybook/blob/master/lib/ui/src/modules/ui/components/layout/index.js#L201) and it [returns a function](https://github.com/storybooks/storybook/blob/master/lib/ui/src/modules/ui/components/layout/index.js#L146) which is ought to handle some event with size as a parameter.

I couldn't find requestAnimationFrame or something similar [in the lib which the sidebar is implemented with](https://github.com/tomkp/react-split-pane/blob/master/src/SplitPane.js#L40), but that's not really a performance bottleneck in my opinion (you can check the timeline of rerenders with `rect_perf` and decide for yourself).

So yup, this issue i guess is solved, given that it didn't existed to begin with, what do you think?

**Comment by wuweiweiwu** [comment]
@shilman @Hypnosphi I just took a peek at the code and im 99.99% sure the issue lies because we dispatch `setState` too many times [here] (https://github.com/storybooks/storybook/blob/master/lib/ui/src/modules/ui/components/layout/index.js#L164)

And MDN even has a [page](https://developer.mozilla.org/en-US/docs/Web/Events/resize) dedicated to how to throttle resize events. 

But I'll have a PR out for this in the next couple of days when I have time :)

**Comment by danielduan** [comment]
In that case, a `throttle` probably makes more sense than `requestAnimationFrame`?

we shouldn't be running `setState` 60x / sec even if the computer is fast enough to support it. Looking forward to your PR, thanks @wuweiweiwu

**Comment by Hypnosphi** [comment]
> we shouldn't be running setState 60x / sec

Why? I'd like the resizing to be smooth

**Comment by wuweiweiwu** [comment]
@Hypnosphi @danielduan Done! I took a slightly different approach and delayed state update until the user stops dragging for 50ms using `setTimeout` and `clearTimeout` so we don't call `setState` a billion times :)

**Comment by Hypnosphi** [comment]
Released as [`3.4.0-rc.1`](https://github.com/storybooks/storybook/releases/tag/v3.4.0-rc.1)
